### ISSUE TITLE
Remove the no-args constructor requirement in Compact Serializer

### ISSUE THREAD

**Issue by TomaszGaweda** [hazelcast/hazelcast#19494]
Currently Compact Serializer requires user to have no-args constructors. While this being standard in entities (like JPA), it's inconvenient when you want to eg. send computation task with some internal state (parameters) or some flags. 
In current Hazelcast with eg. Kryo you can have non-default constructor and it will work just fine, allowing user to archive better encapsulation. 
There's a simple way to do this - use objenesis library. This will require new dependency in core module, but in my humble opinion, it's worth it.

**Comment by TomaszGaweda** [comment]
For the context: it was inspired by this comment: https://github.com/hazelcast/hazelcast/pull/19315#issuecomment-913599181

**Comment by mmedenjak** [comment]
Thank you for the suggestion @TomaszGaweda ! @sancar @mdumandag can you assess whether this is something we should address?

**Comment by mdumandag** [comment]
Well, I think this is something we should do for easier use. Especially, we should remove this constraint for reflective serialization of Java `record`s (you can define no-args constructor for `record`s as well, but I doubt most will do it)

I just checked the produced JAR size after shading the `objenesis` library, and there is nothing extraordinary.

Commit: 06b3d99e1379e7389434230a9d94a7dfc05bf967
Before: 14,9 MB (14.862.580 bytes)
After: 14,9 MB (14.908.700 bytes)

Also, Objenesis is open source, under the Apache 2.0 license. So, there should be no problem there as well.

My feeling is that it should be OK to use `objenesis` as a dependency here to solve the problem.

**Comment by TomaszGaweda** [comment]
IMHO for records I'll create separate PR (for this issue I'll create such this week)

**Comment by mdumandag** [comment]
Thanks for the analysis @jerrinot. I mistakenly thought we can get away with it quickly without looking too much into it :sweat_smile: 

There is definitely extra care needed to handle records and I will put this to backlog

**Comment by sancar** [comment]
Fixed by https://github.com/hazelcast/hazelcast/pull/19752
